### PR TITLE
Code cleanup 

### PR DESCRIPTION
**Pull request by xen**
What I purpose with this PR:

- Code simplification, it is little messed in IDE when you see all methods twice
- PEP8 cleanup
- Switch to mixin type of inheritance

I hope it will help to maintain and develop code in the future.

### CODE REVIEW

File: instabot/bot/bot.py [hunk 1]
```diff
@@ -4,51 +4,28 @@
...
-from .bot_filter import check_not_bot
-
-from .bot_support import check_if_file_exists, read_list_from_file, check_whitelists
-from .bot_support import add_whitelist, add_blacklist
-
-from .bot_stats import save_user_stats
-
 
-class Bot(API):
+class Bot(API, BotArchive, BotBlock, BotComment, BotDelete,
```
Comment (sudoguy) [inline instabot/bot/bot.py:25]: Wrong method resolution order. API must be in end

File: instabot/bot/bot.py [hunk 2]
```diff
@@ -79,9 +56,8 @@ def __init__(self,
                  comment_delay=60,
                  block_delay=30,
                  unblock_delay=30,
-                 stop_words=['shop', 'store', 'free']):
-        super(Bot, self).__init__()
-
+                 stop_words=None):
+        super(self.__class__, self).__init__()
```
Comment (sudoguy) [inline instabot/bot/bot.py:60]: API init

Comment (sudoguy) [reply instabot/bot/bot.py:60]: Not ```self.__class__```

File: instabot/bot/bot_archive.py [hunk 3]
```diff
@@ -3,43 +3,45 @@
...
-        self.total_unarchived += int(undo)
-        return True
-    self.logger.info("Media with %s is not %s ." % media_id, 'unarchived' if undo else 'archived')
-    return False
+class BotArchive(object):
+    def archive(self, media_id, undo=False):
+        delay.small_delay(self)
+        media = self.get_media_info(media_id)
+        media = media[0] if isinstance(media, list) else media
+        if super(self.__class__, self).archiveMedia(media, undo):
```
Comment (sudoguy) [inline instabot/bot/bot_archive.py:11]: API class method

File: instabot/bot/bot_block.py [hunk 4]
```diff
@@ -5,63 +5,61 @@
...
+    def block(self, user_id):
+        user_id = self.convert_to_user_id(user_id)
+        if self.check_not_bot(user_id):
             return True
-    else:
-        self.logger.info("Out of blocks for today.")
-    return False
+        if limits.check_if_bot_can_block(self):
+            delay.block_delay(self)
+            if super(self.__class__, self).block(user_id):
```
Comment (sudoguy) [inline instabot/bot/bot_block.py:15]: API class method

File: instabot/bot/bot_block.py [hunk 5]
```diff
@@ -5,63 +5,61 @@
...
+                return True
+        else:
+            self.logger.info("Out of blocks for today.")
+        return False
 
+    def unblock(self, user_id):
+        user_id = self.convert_to_user_id(user_id)
+        if limits.check_if_bot_can_unblock(self):
+            delay.unblock_delay(self)
+            if super(self.__class__, self).unblock(user_id):
```
Comment (sudoguy) [inline instabot/bot/bot_block.py:26]: API class method

File: instabot/bot/bot_comment.py [hunk 6]
```diff
@@ -14,67 +14,62 @@
...
-        return False
-    return self.comment_medias(medias[:amount])
-
-
-def comment_users(self, user_ids, ncomments=None):
-    for user_id in user_ids:
-        if not limits.check_if_bot_can_comment(self):
+        if limits.check_if_bot_can_comment(self):
+            delay.comment_delay(self)
+            if super(self.__class__, self).comment(media_id, comment_text):
```
Comment (sudoguy) [inline instabot/bot/bot_comment.py:23]: API class method

File: instabot/bot/bot_delete.py [hunk 7]
```diff
@@ -3,26 +3,28 @@
...
-    if super(self.__class__, self).deleteMedia(media):
-        return True
-    self.logger.info("Media with %s is not %s." % (media.get('id'), 'deleted'))
-    return False
+class BotDelete(object):
+    def delete_media(self, media_id):
+        delay.small_delay(self)
+        media = self.get_media_info(media_id)
+        media = media[0] if isinstance(media, list) else media
+        if super(self.__class__, self).deleteMedia(media):
```
Comment (sudoguy) [inline instabot/bot/bot_delete.py:11]: API class method

File: instabot/bot/bot_direct.py [hunk 8]
```diff
@@ -3,140 +3,144 @@
...
+        :param user_ids: list of user_ids for creating group or one user_id for send to one person
+        :param thread_id: thread_id
+        """
+        user_ids = self._get_user_ids(user_ids)
+        if not isinstance(text, str) and type(user_ids) not in (list, str):
+            self.logger.error(
+                'Text must be an string, user_ids must be an list or string')
+            return False
+        delay.small_delay(self)
+        if super(self.__class__, self).sendDirectItem('message', user_ids, text=text, thread=thread_id):
```
Comment (sudoguy) [inline instabot/bot/bot_direct.py:20]: API class method

File: instabot/bot/bot_direct.py [hunk 9]
```diff
@@ -3,140 +3,144 @@
...
+        """
+        user_ids = self._get_user_ids(user_ids)
+        if not isinstance(text, str) and type(user_ids) not in (list, str):
+            self.logger.error(
+                'Text must be an string, user_ids must be an list or string')
+            return False
+        media = self.get_media_info(media_id)
+        media = media[0] if isinstance(media, list) else media
+        delay.small_delay(self)
+        if super(self.__class__, self).sendDirectItem('media_share', user_ids, text=text, thread=thread_id,
```
Comment (sudoguy) [inline instabot/bot/bot_direct.py:56]: API class method

File: instabot/bot/bot_direct.py [hunk 10]
```diff
@@ -3,140 +3,144 @@
...
+        :param user_ids: list of user_ids for creating group or one user_id for send to one person
+        :param thread_id: thread_id
+        """
+        user_ids = self._get_user_ids(user_ids)
+        if not isinstance(text, str) and type(user_ids) not in (list, str):
+            self.logger.error(
+                'Text must be an string, user_ids must be an list or string')
+            return False
+        delay.small_delay(self)
+        if super(self.__class__, self).sendDirectItem('hashtag', user_ids, text=text, thread=thread_id,
```
Comment (sudoguy) [inline instabot/bot/bot_direct.py:91]: API class method

File: instabot/bot/bot_direct.py [hunk 11]
```diff
@@ -3,140 +3,144 @@
...
+        :param thread_id: thread_id
+        """
+        profile_id = self.convert_to_user_id(profile_user_id)
+        user_ids = self._get_user_ids(user_ids)
+        if not isinstance(text, str) and type(user_ids) not in (list, str):
+            self.logger.error(
+                'Text must be an string, user_ids must be an list or string')
+            return False
+        delay.small_delay(self)
+        if super(self.__class__, self).sendDirectItem('profile', user_ids, text=text, thread=thread_id,
```
Comment (sudoguy) [inline instabot/bot/bot_direct.py:114]: API class method

File: instabot/bot/bot_direct.py [hunk 12]
```diff
@@ -3,140 +3,144 @@
...
+        :param user_ids: list of user_ids for creating group or one user_id for send to one person
+        :param thread_id: thread_id
+        """
+        user_ids = self._get_user_ids(user_ids)
+        if type(user_ids) not in (list, str):
+            self.logger.error(
+                'Text must be an string, user_ids must be an list or string')
+            return False
+        delay.small_delay(self)
+        if super(self.__class__, self).sendDirectItem('like', user_ids, thread=thread_id):
```
Comment (sudoguy) [inline instabot/bot/bot_direct.py:135]: API class method

File: instabot/bot/bot_follow.py [hunk 13]
```diff
@@ -4,79 +4,82 @@
...
+        print('\n ===> Going to Follow user_id: %s ' %
+              (user_id))  # Log to console
+        if not self.check_user(user_id):
             return True
-    else:
-        self.logger.info("Out of follows for today.")
-    return False
+        if limits.check_if_bot_can_follow(self):
+            delay.follow_delay(self)
+            if super(self.__class__, self).follow(user_id):
```
Comment (sudoguy) [inline instabot/bot/bot_follow.py:16]: API class method

File: instabot/bot/bot_like.py [hunk 14]
```diff
@@ -3,98 +3,93 @@
...
-            return True
-    else:
-        self.logger.info("Out of likes for today.")
-    return False
-
+class BotLike(object):
+    def like(self, media_id):
+        if limits.check_if_bot_can_like(self):
+            delay.like_delay(self)
+            if super(self.__class__, self).like(media_id):
```
Comment (sudoguy) [inline instabot/bot/bot_like.py:10]: API class method

File: instabot/bot/bot_photo.py [hunk 15]
```diff
@@ -4,40 +4,41 @@
...
-    delay.small_delay(self)
-    if super(self.__class__, self).uploadPhoto(photo, caption, upload_id):
-        self.logger.info("Photo '%s' is %s ." % (photo, 'uploaded'))
-        return True
-    self.logger.info("Photo '%s' is not %s ." % (photo, 'uploaded'))
-    return False
+class BotPhoto(object):
+    def upload_photo(self, photo, caption=None, upload_id=None):
+        delay.small_delay(self)
+        if super(self.__class__, self).uploadPhoto(photo, caption, upload_id):
```
Comment (sudoguy) [inline instabot/bot/bot_photo.py:10]: API class method

File: instabot/bot/bot_photo.py [hunk 16]
```diff
@@ -4,40 +4,41 @@
...
+    def download_photo(self, media_id, path='photos/', filename=None, description=False):
+        delay.small_delay(self)
+        if not os.path.exists(path):
+            os.makedirs(path)
+        if description:
+            media = self.get_media_info(media_id)[0]
+            caption = media['caption']['text']
+            with open('{path}{0}_{1}.txt'.format(media['user']['username'], media_id, path=path), encoding='utf8', mode='w') as f:
+                f.write(caption)
+        photo = super(self.__class__, self).downloadPhoto(
```
Comment (sudoguy) [inline instabot/bot/bot_photo.py:25]: API class method

File: instabot/bot/bot_unfollow.py [hunk 17]
```diff
@@ -4,81 +4,80 @@
...
+    def unfollow(self, user_id):
+        user_id = self.convert_to_user_id(user_id)
+        user_info = self.get_user_info(user_id)
+        print('\n===> Going to UN-Follow user_id: %s , user_name: %s' %
+              (user_id, user_info["username"]))
+        if self.check_user(user_id):
+            return True  # whitelisted user
+        if limits.check_if_bot_can_unfollow(self):
+            delay.unfollow_delay(self)
+            if super(self.__class__, self).unfollow(user_id):
```
Comment (sudoguy) [inline instabot/bot/bot_unfollow.py:17]: API class method

File: instabot/bot/bot_unlike.py [hunk 18]
```diff
@@ -4,31 +4,30 @@
...
-            self.total_unliked += 1
-            return True
-    else:
-        self.logger.info("Out of unlikes for today.")
-    return False
+class BotUnlike(object):
+    def unlike(self, media_id):
+        if limits.check_if_bot_can_unlike(self):
+            delay.unlike_delay(self)
+            if super(self.__class__, self).unlike(media_id):
```
Comment (sudoguy) [inline instabot/bot/bot_unlike.py:11]: API class method

File: instabot/bot/bot_video.py [hunk 19]
```diff
@@ -1,11 +1,12 @@
...
-    if not super(self.__class__, self).uploadVideo(video, thumbnail, caption):
-        self.logger.info("Video '%s' is not %s ." % (video, 'uploaded'))
-        return False
-    self.logger.info("Video '{video}' uploaded".format(video=video))
-    return True
+class BotVideo(object):
+    def upload_video(self, video, thumbnail, caption=''):
+        delay.small_delay(self)
+        self.logger.info("Started uploading '{video}'".format(video=video))
+        if not super(self.__class__, self).uploadVideo(video, thumbnail, caption):
```
Comment (sudoguy) [inline instabot/bot/bot_video.py:8]: API class method

**Review by sudoguy — CHANGES_REQUESTED** [review]
Wrong method resolution order. Method super() can't find right function

**Comment by ohld** [comment]
@xen Will you fix the issues with the super class? Your pull request is awesome but there are several bugs that should be fixed!

**Comment by ohld** [comment]
@xen Waiting for your fixes.

**Pull request closed without merging**